registration/include/pcl/registration/impl/ndt.hpp

```diff
@@ -242,83 +278,139 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeAngleDerivatives
     // Derivative w.r.t. ith and jth elements of transform vector corresponds to the 3x1 block matrix starting at (3i,j), Equation 6.20 and 6.21 [Magnusson 2009]
-    point_hessian_.block<3, 1>(9, 3) = a;
-    point_hessian_.block<3, 1>(12, 3) = b;
-    point_hessian_.block<3, 1>(15, 3) = c;
-    point_hessian_.block<3, 1>(9, 4) = b;
-    point_hessian_.block<3, 1>(12, 4) = d;
-    point_hessian_.block<3, 1>(15, 4) = e;
-    point_hessian_.block<3, 1>(9, 5) = c;
-    point_hessian_.block<3, 1>(12, 5) = e;
-    point_hessian_.block<3, 1>(15, 5) = f;
+    point_hessian.block<4, 1>((9/3)*4, 3) = a;
+    point_hessian.block<4, 1>((12/3)*4, 3) = b;
+    point_hessian.block<4, 1>((15/3)*4, 3) = c;
+    point_hessian.block<4, 1>((9/3)*4, 4) = b;
+    point_hessian.block<4, 1>((12/3)*4, 4) = d;
+    point_hessian.block<4, 1>((15/3)*4, 4) = e;
+    point_hessian.block<4, 1>((9/3)*4, 5) = c;
+    point_hessian.block<4, 1>((12/3)*4, 5) = e;
+    point_hessian.block<4, 1>((15/3)*4, 5) = f;
+  }
+}
+
+// \note Thread-unsafe ```updateDerivatives``` is deprecated
+template<typename PointSource, typename PointTarget> double
+NormalDistributionsTransform<PointSource, PointTarget>::updateDerivatives (Eigen::Matrix<double, 6, 1> &score_gradient,
+                                                                           Eigen::Matrix<double, 6, 6> &hessian,
+                                                                           const Eigen::Vector3d &x_trans, const Eigen::Matrix3d &c_inv,
+                                                                           bool compute_hessian)
+{
+  if (num_threads_ > 1) {
+    PCL_WARN("This implementation is thread unsafe. Please use the thread safe version.");
+  }
+
+  Eigen::Matrix<double, 4, 6> point_jacobian = Eigen::Matrix<double, 4, 6>::Zero();
+  Eigen::Matrix<double, 24, 6> point_hessian = Eigen::Matrix<double, 24, 6>::Zero();
+  point_jacobian.block<3, 6>(0, 0) = point_jacobian_;
+  for (int i = 0; i < 6; i++) {
+    point_hessian.block<3, 6>(i * 4, 0) = point_hessian_.block<3, 6>(i * 3, 0);
   }
+
+  double score_inc = updateDerivatives(score_gradient, hessian, point_jacobian, point_hessian, x_trans, c_inv, compute_hessian);
+
+  point_jacobian_ = point_jacobian.block<3, 6>(0, 0);
+  for (int i = 0; i < 6; i++) {
+    point_hessian_.block<3, 6>(i * 3, 0) = point_hessian.block<3, 6>(i * 4, 0);
+  }
+
+  return score_inc;
 }
 
 
 template<typename PointSource, typename PointTarget> double
 NormalDistributionsTransform<PointSource, PointTarget>::updateDerivatives (Eigen::Matrix<double, 6, 1> &score_gradient,
                                                                            Eigen::Matrix<double, 6, 6> &hessian,
+                                                                           const Eigen::Matrix<double, 4, 6> &point_jacobian,
+                                                                           const Eigen::Matrix<double, 24, 6> &point_hessian,
```

Comment:
Might as well request x_trans and c_inv to be Vector4d.

PS: Please include a note that 4d is needed instead of 3d because of auto-vec

Reply:
Done